xrpl/utils/txn_parser/utils/balance_parser.py

```diff
@@ -0,0 +1,185 @@
+"""Helper functions for balance parser."""
+
+from decimal import Decimal
+from itertools import chain
+from typing import Dict, List, Optional, Union
+
+from xrpl.utils.txn_parser.utils.nodes import NormalizedNode
+from xrpl.utils.txn_parser.utils.types import Balance, BalanceChange, BalanceChanges
+from xrpl.utils.xrp_conversions import drops_to_xrp
+
+
+def _get_xrp_quantity(
+    node: NormalizedNode,
+    value: Optional[Decimal],
+) -> Optional[BalanceChange]:
+    if value is None:
+        return None
+    absolute_value = value.copy_abs()
+    xrp_value = (
+        drops_to_xrp(str(absolute_value)).copy_negate()
+        if value.is_signed()
+        else drops_to_xrp(str(absolute_value))
+    )
+    final_fields = node.get("FinalFields")
+    if final_fields is not None:
+        account = final_fields.get("Account")
+        if account is not None:
+            return BalanceChange(
+                account=account,
+                balance=Balance(  # type: ignore
+                    currency="XRP",
+                    value=f"{xrp_value.normalize():f}",
+                ),
+            )
+    new_fields = node.get("NewFields")
+    if new_fields is not None:
+        account = new_fields.get("Account")
+        if account is not None:
+            return BalanceChange(
+                account=account,
+                balance=Balance(  # type: ignore
+                    currency="XRP",
+                    value=f"{xrp_value.normalize():f}",
+                ),
+            )
+    return None
+
+
+def _flip_trustline_perspective(balance_change: BalanceChange) -> BalanceChange:
+    balance = balance_change["balance"]
+    negated_value = Decimal(balance["value"]).copy_negate()
+    issuer = balance["issuer"]
+    return BalanceChange(
+        account=issuer,  # type: ignore
```

Comment:
Same question

Reply:
`issuer` is str or None. I added `cast` and removed `# type: ignore`.

Reply:
It looks like the way that `TypedDict` works, you should omit the optional variables instead of marking them as `Optional`.

Reply:
I'd recommend just leaving a comment in the `Balance` `TypedDict` that there's an optional key, `issuer`

Reply:
The `Balance` `TypedDict` now looks like this:
```python
class Balance(TypedDict):
    """A account's balance model."""

    currency: str
    """The currency."""
    issuer: str
    """The issuer of the currency. This value is optional."""
    value: str
    """The amount of the currency."""
```
Now I do not have to use `cast` anymore to pass the `issuer` in the `Balance` `TypedDict`. But the two problems above still remain. mypy still raises: Missing key "issuer" for TypedDict "Balance"

Reply:
If you make it look like this:
```py
class Balance(TypedDict):
    """A account's balance model."""

    currency: str
    """The currency."""
    value: str
    """The amount of the currency."""
    # issuer: str
    # """The issuer of the currency. This value is optional."""
```
I think that should work

Reply:
I found a way that seems to work fine:
```python
Balance = TypedDict(
    "Balance", {"currency": str, "issuer": str, "value": str}, total=False
)
```

Reply:
This is the best solution I could come up with:
```python
class IssuerField(TypedDict, total=False):
    """The optional `issuer` field for an account's balance."""

    issuer: str
    """The issuer of the currency. This value is optional."""


class Balance(IssuerField):
    """A account's balance model."""

    currency: str
    """The currency code."""
    value: str
    """The amount of the currency."""
```
It works perfectly fine. Maybe `IssuerField` might not be the best name for the `TypedDict`. Please give me a short feedback :)

Reply:
One last idea is to use `NotRequired[...]` from `typing_extensions` instead of `Optional[...]` (See answer with it in: https://stackoverflow.com/questions/68814891/typeddict-does-not-allow-optional-keys)

Note: You have to use TypedDict from `typing_extensions` since this is a very new feature which isn't in normal `typing` until Python 3.11, and we support back to Python 3.7. 

If that doesn't work, I think the solution you put here works with two small changes:
1. I would add a one-line comment above the `IssuerField` explaining why it's split into two
2. Maybe `OptionalIssuer` would be a better name than `IssuerField`?

Reply:
I already tried `NotRequired`. Unfortunately it has no affect on mypy. It still raises: Missing key "issuer"...

Reply:
Ah ok, well then I think your solution is solid :)

Reply:
Do you think it would make sense to do the same for all other optional values in the TypedDicts (e.g. `NormalizedNode`, `TransactionMetadata`, `ModifiedNodeFields`)? Or do you think this is not necessary?

Reply:
> If you can figure out how to remove that additional `type: ignore`, that would be great

I found a way to remove the last `type: ignore`, but I find it a bit messy. Maybe you can come up with an easier way, or simplify mine.